gateway-service/src/main/java/org/zowe/apiml/gateway/security/service/token/OIDCTokenProvider.java

```diff
@@ -10,72 +10,125 @@
+import org.apache.http.HttpStatus;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.entity.UrlEncodedFormEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.message.BasicNameValuePair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.beans.factory.annotation.Value;
-import org.springframework.http.HttpEntity;
 import org.springframework.http.HttpHeaders;
-import org.springframework.http.HttpMethod;
-import org.springframework.http.ResponseEntity;
+import org.springframework.http.MediaType;
 import org.springframework.stereotype.Service;
-import org.springframework.web.client.RestClientException;
-import org.springframework.web.client.RestTemplate;
-import org.zowe.apiml.gateway.security.service.schema.OIDCAuthException;
 import org.zowe.apiml.security.common.token.OIDCProvider;
 
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
 import java.util.Base64;
+import java.util.List;
 
 @RequiredArgsConstructor
 @Service
 @Slf4j
 public class OIDCTokenProvider implements OIDCProvider {
 
-    @Value("${apiml.security.oAuth.clientId:}")
+    @Value("${apiml.security.oidc.introspectEndpoint:/introspect")
+    private String introspectEndpoint;
+
+    @Value("${apiml.security.oidc.clientId:}")
     private String clientId;
 
-    @Value("${apiml.security.oAuth.clientSecret:}")
+    @Value("${apiml.security.oidc.clientSecret:}")
     private String clientSecret;
 
-    @Value("${apiml.security.oAuth.validationUrl:}")
-    private String validationUrl;
-
-    @Value("${apiml.security.oAuth.enabled:false}")
+    @Value("${apiml.security.oidc.enabled:false}")
     private boolean isEnabled;
 
-    private final RestTemplate restTemplate;
+    @Autowired
+    @Qualifier("secureHttpClientWithoutKeystore")
+    @NonNull
+    private final CloseableHttpClient httpClient;
+
+    private static final ObjectMapper mapper = new ObjectMapper();
 
     @Override
-    public boolean isValid(String token) {
-        if (token == null || token.isEmpty() || !isEnabled) {
-            log.debug("Either you did not enable the OIDC auth or you did not provide a valid token.");
-            throw new OIDCAuthException("A failure occurred when validating.");
+    public boolean isValid(String token, String issuer) {
+        OIDCTokenClaims claims = introspect(token, issuer);
+        if (claims != null) {
+            return claims.getActive();
         }
+        return false;
+    }
+
+    private OIDCTokenClaims introspect(String token, String issuer) {
+        if (!isEnabled) {
```

Comment:
can this happen? would be better if there is no bean at all if it is not enabled

Reply:
Bean is created conditionally if oidc is enabled.